src/model_theory/types.lean

```diff
@@ -151,6 +156,52 @@ end
 
 end complete_type
 
+variables {M : Type w'} [L.Structure M] [nonempty M] [M ⊨ T] (T)
+
+/-- The set of all formulas true at a tuple in a structure forms a complete type. -/
+def type_of (v : α → M) : T.complete_type α :=
+begin
+  haveI : (constants_on α).Structure M := constants_on.Structure v,
+  exact { to_Theory := L[[α]].complete_theory M,
+    subset' := model_iff_subset_complete_theory.1 ((Lhom.on_Theory_model _ T).2 infer_instance),
+    is_maximal' := complete_theory.is_maximal _ _ },
+end
+
+namespace complete_type
+
+variables {T} {v : α → M}
+
+@[simp] lemma mem_type_of {φ : L[[α]].sentence} :
+  φ ∈ T.type_of v ↔ (formula.equiv_sentence.symm φ).realize v :=
+begin
+  letI : (constants_on α).Structure M := constants_on.Structure v,
+  exact mem_complete_theory.trans (formula.realize_equiv_sentence_symm _ _ _).symm,
+end
+
+lemma formula_mem_type_of {φ : L.formula α} :
+  (formula.equiv_sentence φ) ∈ T.type_of v ↔ φ.realize v :=
+by simp
+
+/-- A complete type `p` is realized in a particular structure when there is some
+  tuple `v` whose type is `p`. -/
+def is_realized_in (p : T.complete_type α) (M : Type w') [L.Structure M] [nonempty M] [M ⊨ T] :
+  Prop :=
+∃ v : α → M, T.type_of v = p
```

Comment:
This is the preimage of `{p}` under `T.type_of`. Are there preimage lemmas that you can transfer to this new definition?

Reply:
I don't see that - but if you fix a particular `M`, isn't it the _image_ of `T.type_of`?
The naming is a bit trickier then, but it is useful to think of that image set topologically, so perhaps I should pivot to that variable order...

Reply:
Sorry, that was a late night comment.🤦 Of course this is the range.